File: emisor/capas/transmision.py
"""Capa de transmision: serializa/deserializa el sobre JSON sobre el socket TCP."""

import json
import socket


class LectorLineas:
    """Acumula bytes del socket hasta encontrar '\\n', porque TCP puede partir o pegar mensajes."""

    def __init__(self, sock: socket.socket):
        self._sock = sock
        self._buffer = b""

    def leer_linea(self) -> str | None:
        while b"\n" not in self._buffer:
            fragmento = self._sock.recv(4096)
            if not fragmento:
                return None
            self._buffer += fragmento
        linea, _, self._buffer = self._buffer.partition(b"\n")
        return linea.decode("utf-8")


def enviar_informacion(sock: socket.socket, trama: str, algoritmo: str, parametros: dict) -> None:
    """Serializa el sobre JSON y lo envia terminado en salto de linea."""
    sobre = {"version": 1, "algoritmo": algoritmo, "parametros": parametros, "trama": trama}
    sock.sendall((json.dumps(sobre) + "\n").encode("utf-8"))


def recibir_informacion(lector: LectorLineas) -> dict | None:
    """Lee hasta el salto de linea y devuelve el sobre deserializado, o None si se cerro la conexion."""
    linea = lector.leer_linea()
    if linea is None:
        return None
    try:
        sobre = json.loads(linea)
    except json.JSONDecodeError as error:
        raise ValueError("sobre JSON invalido") from error
    if not isinstance(sobre, dict) or sobre.get("version") != 1:
        raise ValueError("version de sobre no soportada")
    if not isinstance(sobre.get("algoritmo"), str) or not isinstance(sobre.get("parametros"), dict):
        raise ValueError("campos basicos del sobre invalidos")
    trama = sobre.get("trama")
    if not isinstance(trama, str) or any(bit not in "01" for bit in trama):
        raise ValueError("trama debe ser una cadena binaria")
    return sobre

File: receptor/capas/transmision.js
/** Capa de transmision: serializa/deserializa el sobre JSON sobre el socket TCP. Sin capa de ruido. */

/** Acumula fragmentos hasta encontrar '\n', porque TCP puede partir o pegar mensajes. */
export class LectorLineas {
  constructor() {
    this._buffer = "";
  }

  /** @param {string} fragmento */
  agregar(fragmento) {
    this._buffer += fragmento;
  }

  /** @returns {Generator<string>} lineas completas disponibles en el buffer, en orden */
  *extraerLineas() {
    let indice = this._buffer.indexOf("\n");
    while (indice !== -1) {
      yield this._buffer.slice(0, indice);
      this._buffer = this._buffer.slice(indice + 1);
      indice = this._buffer.indexOf("\n");
    }
  }
}

/** @param {import("node:net").Socket} conn @param {string} trama @param {string} algoritmo @param {Object} parametros */
export function enviarInformacion(conn, trama, algoritmo, parametros) {
  const sobre = { version: 1, algoritmo, parametros, trama };
  conn.write(`${JSON.stringify(sobre)}\n`);
}

/** @param {string} linea @returns {Object} */
export function recibirInformacion(linea) {
  let sobre;
  try { sobre = JSON.parse(linea); } catch { throw new Error("sobre JSON invalido"); }
  if (!sobre || sobre.version !== 1) throw new Error("version de sobre no soportada");
  if (typeof sobre.algoritmo !== "string" || typeof sobre.parametros !== "object") throw new Error("campos basicos del sobre invalidos");
  if (typeof sobre.trama !== "string" || /[^01]/.test(sobre.trama)) throw new Error("trama debe ser una cadena binaria");
  return sobre;
}
